Feat/format conditions multi line

Syzygy106 opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/foundry-rs/foundry/blob/HEAD/CONTRIBUTING.md

The contributors guide includes instructions for running rustfmt and building the
documentation.
-->

<!-- ** Please select "Allow edits from maintainers" in the PR Options ** -->

## Motivation

<!--
Explain the context and why you're making that change. What is the problem
you're trying to solve? In some cases there is not a problem and this can be
thought of as being the motivation for your change.
-->

Closes #12508

When conditions in `if` statements are formatted inline (all on one line), coverage analysis tools like lcov cannot properly distinguish which individual conditions within the chain are actually covered by tests. This happens because the tool sees only one line with the condition, making it difficult to determine branch coverage for each condition.

For example:
```solidity
if (newNumber % 2 == 0 || newNumber % 2 == 1 || newNumber != 0 || newNumber != 1 || newNumber != 2) {
    number = newNumber;
}
```

With inline formatting, if a test only covers the first condition (e.g., `newNumber % 2 == 0`), lcov may show the entire line as covered, but it cannot indicate which specific conditions were actually tested. This makes it difficult to identify untested conditions and achieve accurate branch coverage analysis.



## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

Added a new `format_conditions` configuration option in `foundry.toml` that allows placing each condition on a separate line:

```toml
[fmt]
format_conditions = "multi"  # or "inline" (default)
```

When `format_conditions = "multi"`, logical operators (`||`, `&&`) in `if` conditions are formatted with each condition on a separate line:

```solidity
if (
    newNumber % 2 == 0
        ||
    newNumber % 2 == 1
        ||
    newNumber != 0
        ||
    newNumber != 1
        ||
    newNumber != 2
) {
    number = newNumber;
}
```

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [ ] Breaking changes

Syzygy106 committed:
feat(fmt): add format_conditions config option

Syzygy106 committed:
feat(fmt): multi-line condition formatting

Syzygy106 commented:
```
======================================================================

1. With format_conditions = 'inline' (default):

// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

contract Counter {
    uint256 public number = 1;

    function setNumber(uint256 newNumber) public {
        if (newNumber % 2 == 0 || newNumber % 2 == 1 || newNumber != 0 || newNumber != 1 || newNumber != 2) {
            number = newNumber;
        }
    }

    function increment() public {
        number++;
    }
}


======================================================================

2. With format_conditions = 'multi':

// SPDX-License-Identifier: UNLICENSED
pragma solidity ^0.8.13;

contract Counter {
    uint256 public number = 1;

    function setNumber(uint256 newNumber) public {
        if (
            newNumber % 2 == 0
                ||
                    newNumber % 2 == 1
                ||
                    newNumber != 0
                ||
                    newNumber != 1
                ||
                    newNumber != 2
        ) {
            number = newNumber;
        }
    }

    function increment() public {
        number++;
    }
}
```

Syzygy106 committed:
fix(tests): added missing parameter

grandizzy reviewed (dismissed):
thank you, makes sense IMO (pending @0xrusowsky ). Please update the readme too to reflect the new config https://github.com/foundry-rs/foundry/tree/master/crates/fmt#configuration and add an ui_test as in https://github.com/foundry-rs/foundry/tree/master/crates/fmt#testing

0xrusowsky commented:
the impl LGTM, but imo it would be nice to make the output prettier.

if possible, let's aim for:
```sol
if (
    newNumber % 2 == 0
        || newNumber % 2 == 1
        || newNumber != 0
        || newNumber != 1
        || newNumber != 2
) {
```

i think you should be able to achieve if with an extra box when the multiline config is enabled

Syzygy106 committed:
feat(fmt): multiline correction + tests + doc

Syzygy106 committed:
chore(fmt): clippy fix

Syzygy106 committed:
chore(fmt): fix convenience formatting with \n

0xrusowsky commented on crates/fmt/src/pp/convenience.rs:
```diff
@@ -31,6 +31,17 @@ impl Printer {
+            while self.out.ends_with("\n\n") {
+                self.out.pop();
+            }
+        }
```
why do we need to introduce this logic?

0xrusowsky replied:
@Syzygy106 can we bring this over the finish line?

gakonst committed:
fix(fmt): remove eof trailing newline workaround

Address review comment from @0xrusowsky: the trailing newline normalization
in eof() was a workaround for a test file issue (original.sol had CRLF line
endings and a trailing blank line).

The proper fix is to clean up the test file instead of post-processing in eof().
This removes the unnecessary normalization logic and fixes the root cause.

Amp-Thread-ID: https://ampcode.com/threads/T-019bdc92-a774-749b-a831-7d9de5021e45
Co-authored-by: Amp <amp@ampcode.com>

DaniPopes committed:
Merge branch 'master' into feat/format-conditions-multi-line

grandizzy reviewed (dismissed):
think this should be good to merge, @0xrusowsky @DaniPopes mind to have a 2nd look? ty!

onbjerg committed:
Merge branch 'master' into feat/format-conditions-multi-line

onbjerg committed:
Merge branch 'master' into feat/format-conditions-multi-line

emmajam commented:
hey we are doing some spring cleaning and closing old pull requests, please reopen this if you think its still needed and valuable, thank you for your contributions! ❤️

Syzygy106 commented:
@grandizzy , @0xrusowsky we almost merged that. What do you think? Could we reopen and merge?
